numba/np/arrayobj.py

```diff
@@ -4665,52 +4666,125 @@ def impl(v, k=0):
-
-        def take_impl(a, indices):
-            ret = np.empty(indices.size, dtype=a.dtype)
-            if F_order:
-                walker = indices.copy()  # get C order
-            else:
-                walker = indices
-            it = np.nditer(walker)
-            i = 0
-            flat = a.ravel()
-            for x in it:
-                if x > (a.size - 1) or x < -a.size:
-                    raise IndexError("Index out of bounds")
-                ret[i] = flat[x]
-                i = i + 1
-            return ret.reshape(indices.shape)
-        return take_impl
-
-    if isinstance(a, types.Array) and \
-            isinstance(indices, (types.List, types.BaseTuple)):
-        def take_impl(a, indices):
-            convert = np.array(indices)
-            ret = np.empty(convert.size, dtype=a.dtype)
-            it = np.nditer(convert)
-            i = 0
-            flat = a.ravel()
-            for x in it:
-                if x > (a.size - 1) or x < -a.size:
+def numpy_take(a, indices, axis=None):
+
+    if cgutils.is_nonelike(axis):
+        if isinstance(a, types.Array) and isinstance(indices, types.Integer):
+            def take_impl(a, indices, axis=None):
+                if indices > (a.size - 1) or indices < -a.size:
                     raise IndexError("Index out of bounds")
-                ret[i] = flat[x]
-                i = i + 1
-            return ret.reshape(convert.shape)
-        return take_impl
+                return a.ravel()[indices]
+            return take_impl
+
+        if isinstance(a, types.Array) and isinstance(indices, types.Array):
+            F_order = indices.layout == 'F'
+
+            def take_impl(a, indices, axis=None):
+                ret = np.empty(indices.size, dtype=a.dtype)
+                if F_order:
+                    walker = indices.copy()  # get C order
+                else:
+                    walker = indices
+                it = np.nditer(walker)
+                i = 0
+                flat = a.ravel()
+                for x in it:
+                    if x > (a.size - 1) or x < -a.size:
+                        raise IndexError("Index out of bounds")
+                    ret[i] = flat[x]
+                    i = i + 1
+                return ret.reshape(indices.shape)
+            return take_impl
+
+        if isinstance(a, types.Array) and \
+                isinstance(indices, (types.List, types.BaseTuple)):
+            def take_impl(a, indices, axis=None):
+                convert = np.array(indices)
+                return np.take(a, convert)
+            return take_impl
+    else:
+        if isinstance(a, types.Array) and isinstance(indices, types.Integer):
+            t = (0,) * (a.ndim - 1)
+
+            @register_jitable
+            def _squeeze(r):
+                # np.squeeze is too hard to implement in Numba
+                tup = tuple(t)
+                j = 0
+                for s in r.shape:
+                    if s != 1:
```

Comment:
The squeeze implementation needs to exclude the axis based on the axis provided not the value `1`:

For instance:
```python
import time
from numba import njit
import numpy as np

@njit
def foo(a, indices, axis):
    return np.take(a, indices, axis)


def measure_time(fn, *args):
    start = time.time()
    res = fn(*args)
    end = time.time()
    print(f'{fn.__name__}: {end - start}s')
    print(res.shape)
    print("ndim:", res.ndim)



shape = (10, 1, 11, 1, 12, 1, 13)
a = np.ones(shape)
indices = 0
axis=1
measure_time(foo.py_func, a, indices, axis)
# foo: 0.00011563301086425781s
# (10, 11, 1, 12, 1, 13)
# ndim: 6
measure_time(foo, a, indices, axis) # fails

shape = (10, 1, 11, 1, 12, 1, 13)
a = np.ones(shape)
indices = 0
axis = 2
measure_time(foo.py_func, a, indices, axis)
# foo: 6.4373016357421875e-06s
# (10, 1, 1, 12, 1, 13)
# ndim: 6
measure_time(foo, a, indices, axis) # fails

```

Reply:
Good catch. I'll fix this